### PR TITLE
AP_Proximity: Add support for MR72 CAN

### PR DESCRIPTION
**Pull request by rishabsingh3003**
This adds support for the CAN variant of MR72. 
http://en.nanoradar.cn/Article/detail/id/488.html

### CODE REVIEW

**Review by tridge — APPROVED** [review]

**Comment by rishabsingh3003** [comment]
This has now been well tested by me and NanoRadar

File: libraries/AP_Proximity/AP_Proximity.cpp [hunk 1]
```diff
@@ -237,6 +263,15 @@ void AP_Proximity::init()
...
         // initialise status
         state[instance].status = Status::NotConnected;
+
+        // if the backend has some local parameters then make those available in the tree
+        if (drivers[instance] && state[instance].var_info) {
+            backend_var_info[instance] = state[instance].var_info;
+            AP_Param::load_object_from_eeprom(drivers[instance], backend_var_info[instance]);
+
+            // param count could have changed
+            AP_Param::invalidate_count();
```
Comment (IamPete1) [inline libraries/AP_Proximity/AP_Proximity.cpp:273]: `load_object_from_eeprom` already does the invalidate_count.

File: libraries/AP_Proximity/AP_Proximity_MR72_CAN.cpp [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+{
+    if (_current_object_index >= _object_count) {
+        // should never happen
+        return false;
+    }
+    _current_object_index++;
+
+    Vector2f obstacle_fr;
+    obstacle_fr.x = ((frame.data[2] & 0x07U) * 256 + frame.data[3]) * 0.2 - 204.6;
+    obstacle_fr.y = (frame.data[1] * 32 + (frame.data[2] >> 3)) * 0.2 - 500;
```
Comment (IamPete1) [inline libraries/AP_Proximity/AP_Proximity_MR72_CAN.cpp:119]: This is a but odd, I think it deserves a comment.

Comment (rishabsingh3003) [reply libraries/AP_Proximity/AP_Proximity_MR72_CAN.cpp:119]: @IamPete1 this parsing is directly coming from their datasheet. Not sure what I can clarify in the comments to be  honest

Comment (IamPete1) [reply libraries/AP_Proximity/AP_Proximity_MR72_CAN.cpp:120]: Saying look at the data sheet is find, bonus points for also putting the datasheet in our new repo, https://github.com/ArduPilot/Datasheets and linking to it from the comment.

File: libraries/AP_Proximity/AP_Proximity_Params.cpp [hunk 3]
```diff
@@ -8,7 +8,7 @@ const AP_Param::GroupInfo AP_Proximity_Params::var_info[] = {
     // @Param: _TYPE
     // @DisplayName: Proximity type
     // @Description: What type of proximity sensor is connected
-    // @Values: 0:None,7:LightwareSF40c,2:MAVLink,3:TeraRangerTower,4:RangeFinder,5:RPLidarA2,6:TeraRangerTowerEvo,8:LightwareSF45B,10:SITL,12:AirSimSITL,13:CygbotD1, 14:DroneCAN, 15:Scripting, 16:LD06
+    // @Values: 0:None,7:LightwareSF40c,2:MAVLink,3:TeraRangerTower,4:RangeFinder,5:RPLidarA2,6:TeraRangerTowerEvo,8:LightwareSF45B,10:SITL,12:AirSimSITL,13:CygbotD1, 14:DroneCAN, 15:Scripting, 16:LD06, 17: MR72_CAN
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity_Params.cpp:11]: nitpick: no need for the space after the ":".   not a blocker, sorry to be annoying, just my OCD.

File: libraries/AP_Proximity/AP_Proximity_MR72_CAN.cpp [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+{
+    if (_current_object_index >= _object_count) {
+        // should never happen
+        return false;
+    }
+    _current_object_index++;
+
+    Vector2f obstacle_fr;
+    // This parsing comes from the NanoRadar MR72 datasheet
+    obstacle_fr.x = ((frame.data[2] & 0x07U) * 256 + frame.data[3]) * 0.2 - 204.6;
```
Comment (Ryanf55) [inline libraries/AP_Proximity/AP_Proximity_MR72_CAN.cpp:119]: Is there a segfault risk if the CANFrame is shorter length than expected?

File: libraries/AP_Proximity/AP_Proximity_MR72_CAN.cpp [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+#if AP_PROXIMITY_MR72_ENABLED
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_BoardConfig/AP_BoardConfig.h>
+#include "AP_Proximity_MR72_CAN.h"
+
+const AP_Param::GroupInfo AP_Proximity_MR72_CAN::var_info[] = {
+
+    // @Param: RECV_ID
+    // @DisplayName: CAN receive ID
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity_MR72_CAN.cpp:26]: Just regarding naming, "receive" is in the eye of the beholder -- one person's receive is another person's send so it doesn't add much clarity.  maybe call it CAN_ID or just "ID".

**Pull request merged**